libr/core/cmd_type.c

```diff
@@ -341,24 +355,68 @@ static int cmd_type(void *data, const char *input) {
 		case '?': {
 			const char *help_message[] = {
 				"Usage:", "", "",
-				"tl", "<typename>", "link a type to current adress.",
-				"tl", "<typename> <address>", "link type to given address.",
+				"tl", "", "list all links in readable format",
+				"tl", "[typename]", "link a type to current adress.",
+				"tl", "[typename] = [address]", "link type to given address.",
+				"tls", "[address]", "show link at given address",
 				"tl-*", "", "delete all links.",
-				"tl-", "<address>", "delete link at given address.",
+				"tl-", "[address]", "delete link at given address.",
 				"tl*", "", "list all links in radare2 command format",
 				"tl?", "", "print this help.",
 				NULL };
 			r_core_cmd_help (core, help_message);
 			} break;
 		case ' ': {
-			const char *type = input + 2;
-			char *ptr = strchr (type, ' ');
+			char *type = strdup (input + 2);
+			char *ptr = strchr (type, '=');
 			ut64 addr;
+
 			if (ptr) {
 				*ptr++ = 0;
-				addr = r_num_math (core->num, ptr);
-			} else addr = core->offset;
-			r_anal_type_link (core->anal, type, addr);
+				r_str_chop (ptr);
+				if (ptr && *ptr) {
+					addr = r_num_math (core->num, ptr);
+				} else {
+					eprintf ("address is unvalid\n");
+					free (type);
+					break;
+				}
+			} else {
+				addr = core->offset;
+			}
+			r_str_chop (type);
+			char *tmp = sdb_get (core->anal->sdb_types, type, 0);
+			if (tmp && *tmp) {
+				r_anal_type_link (core->anal, type, addr);
+				free (tmp);
+			} else {
+				eprintf ("unknown type %s\n", type);
+			}
+			free (type);
+			}
+			break;
+		case 's': {
+			int ptr;
+			char *addr = strdup (input+2);
+			SdbKv *kv;
+			SdbListIter *sdb_iter;
+			SdbList *sdb_list = sdb_foreach_list (core->anal->sdb_types);
```

Comment:
if you want performance dont use sdb_foreach_list. there's a callback-based way to iterate over an sdb (sdb_foreach()) that doesnt implies allocating the whole db into a new list that will be destroyed later